### ISSUE TITLE
Discrepancies in Reproducing Reported Accuracy and Throughput for MoH on LIBERO Benchmarks

### ISSUE THREAD

**Issue by YilinGao-SHU** [Timsty1/MixtureOfHorizons#2]
# Discrepancies in Reproducing Reported Accuracy and Throughput for MoH on LIBERO Benchmarks

Dear authors,

First of all, thank you for open-sourcing your work—this greatly facilitates reproducibility and community engagement.

I have made a concerted effort to reproduce the results reported in your paper using the official checkpoint (`pi05_moh`) and strictly following the instructions provided in the repository. Despite extensive experimentation—including systematic sweeps over the `scale_ratio` hyperparameter (from 1.1 to 1.4)—I consistently observe two significant discrepancies between the claimed performance and my empirical findings:

---

## 1. Accuracy Gap

Across all tested configurations, the reproduced success rates fall short of the paper’s reported **~99% average on LIBERO** by **1–3 percentage points**. Notably, the largest gap appears on **LIBERO-Long**, where MoH is theoretically expected to excel due to its "long-term foresight" capability. However, I observe success rates of only **~96%**, substantially below the claimed **98.4%**.

A summary of key results (using the best-performing `scale_ratio=1.2`) is as follows:

| Task Suite        | Paper (Reported) | Reproduced (`ratio=1.2`) | Reproduced (`ratio=1.1`) | Gap                |
|-------------------|------------------|--------------------------|--------------------------|--------------------|
| LIBERO-Spatial    | 98.8%            | 98.0%                    | 97.0%                    | -0.8%              |
| LIBERO-Object     | 100.0%           | —                        | 99.4%                    | -0.6%              |
| LIBERO-Goal       | 98.8%            | —                        | 98.2%                    | -0.6%              |
| LIBERO-Long       | 98.4%            | 95.8%                    | 96.2%                    | -2.0% to -2.6%     |
| **Average**       | **99.0%**        | —                        | **97.7%**                | **FAIL**           |

> **Note**: `scale_ratio=1.3` performed particularly poorly (6 failures in the first 150 episodes), while `1.2` yielded the best—but still subpar—results.

---

## 2. Throughput and Latency Mismatch

The paper claims **“minimal overhead”** and **“2.5× higher throughput”** for MoH. However, on an **NVIDIA RTX 4090**, I observe the opposite:

| Model Configuration         | Avg. Latency (ms) | Min/Max Latency (ms) | Throughput (infs/sec) | Avg. Inference Time (s) |
|----------------------------|-------------------|----------------------|------------------------|--------------------------|
| Baseline (`pi05`)          | 79.9              | 78.98 / 83.03        | 12.52                  | 9.29                     |
| MoH (Fixed Horizon)        | 231.28            | 225.95 / 388.76      | 4.32                   | 48.03                    |
| MoH (Dynamic, `r=1.2`)     | 232.25            | 227.00 / 236.09      | 4.31                   | 25.05                    |

**Key observations:**
- **3× higher latency**: MoH inference takes ~232 ms vs. ~80 ms for the baseline.
- **~65% lower throughput**: Throughput drops from 12.52 Hz to ~4.31 Hz.
- **No benefit from dynamic routing**: The dynamic variant shows no throughput advantage over the fixed-horizon version, suggesting that any saved steps are offset by additional computational overhead.

---

## 3. Reproduction Details

To rule out script-level errors, I manually decomposed `scripts/eval_on_libero_dynamic.sh` and explicitly specified dense horizons matching the checkpoint’s training configuration. Below is my full reproduction pipeline:

```bash
# Environment setup
git clone git@github.com:Timsty1/MixtureOfHorizons.git
conda create -n moh -y python=3.10 && conda activate moh
pip install uv
cd MixtureOfHorizons
uv pip install -r requirements.txt
pip install packages/libero packages/openpi-client

# Patch transformers
cp -r ./src/openpi/models_pytorch/transformers_replace/* $CONDA_PREFIX/lib/python3.10/site-packages/transformers/

# Environment variables
export PYTHONPATH=$(pwd)/src:$(pwd)/packages/libero
export MUJOCO_GL=egl
export TORCHDYNAMO_DISABLE=1

# Eval examples (tried all three variants)
python examples/libero_scripts/pi_moh_dynamic.py \
    --task_suite_name="libero_spatial" \
    --rank=0 \
    --video_out_path="videos_dynamic/final_test" \
    --config="pi05_libero" \
    --checkpoint_dir="/data/VLA/checkpoint/pi05_moh" \
    --horizons 3 6 9 12 15 18 21 24 27 30 \
    --scale_ratio=1.2

# Also tested without --horizons and with static pi_moh.py

Logs as below:

[final_test_log.txt](https://github.com/user-attachments/files/24119060/final_test_log.txt)

**Comment by Timsty1** [comment]
Hello, apologies for the delay in seeing this issue. 

First of all, thank you very much for your interest in our work and for using it.

### About Accuracy

The set of results reporting 99% average accuracy **did not use dynamic inference**. Instead, similar to other baseline methods, we used the **first 5 steps** of each predicted action chunk for execution. Please refer to `scripts/eval_on_libero.sh` for reproduction.

We have also open-sourced our test logs and videos; please refer to our [[Google Drive](https://drive.google.com/file/d/1nvuvkfUhu788zyrrN1SZLdKWeN9TZnxs/view)].

While dynamic inference does accelerate the execution process, it indeed causes a certain degree of accuracy drop.

### About Latency

The latency you observed seems a bit unusual.🤔
I suspect there might be some issues with your implementation. I guess that maybe the data parallel does not works well.

Perhaps you could first compare the time consumption of **pi05** vs. **pi05 with MoH** during Libero testing? Based on our experience, there shouldn't be a significant difference between the two. Our experiments were all trained and tested on A100 GPUs, and the reported latency figures are real measured results. For instance, on spatial tasks, both take around **8.5s**. You can verify this in our **logs** as well. 

Additionally, "2.5× higher throughput" refers to the effective usage of actions within the predicted action chunk increasing by 2.5x when the `scale_ratio` is large. However, considering that executing actions in the environment also takes time, the reduction in total inference time might not fully reach 2.5x. That said, as shown in your table, it is already close to 2x.

I hope this answers your questions.
